crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -753,15 +756,35 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn infer_augmented_assignment_statement(&mut self, assignment: &ast::StmtAugAssign) {
-        // TODO this should be a Definition
+        if assignment.target.is_name_expr() {
+            self.infer_definition(assignment);
+        } else {
+            // TODO currently we don't consider assignments to non-Names to be Definitions
+            self.infer_augment_assignment(assignment);
+        }
+    }
+
+    fn infer_augment_assignment_definition(
+        &mut self,
+        assignment: &ast::StmtAugAssign,
+        definition: Definition<'db>,
+    ) {
+        let target_ty = self.infer_augment_assignment(assignment);
+        self.types.definitions.insert(definition, target_ty);
```

Comment:
I think what you're discussing here is orthogonal to representing Unbound as a type or not. In terms of the types that we use for type checking (the types we are confident in), there's no question or issue here: if `x` was unbound before `x += 1`, its still unbound after it (or maybe we prefer switching to unknown in order to avoid more cascading errors, either way can work.)

What you're bringing up is a separate feature, which we discussed in planning this morning, which is that in some cases we may have unknown/any as the "correct" type, but we may also want to track a "best guess" type, so we can give "better" (well, hopefully better rather than just misleading) autocomplete etc downstream of incorrect code. That's a feature that needs some design work and discussion, but I don't think representing Unbound as a type makes it any more difficult to implement that feature.

Reply:
To be clear, I'm very open to considering alternate ways of handling Unbound; I think a good case can be made for other approaches. I just don't think this situation exposes a good rationale for a different approach.

Reply:
I think I need to study this more and propose a concrete alternative. I think mypy's approach to this problem is very different; but then, mypy is in general much worse at detecting unbound variables than either pyright or Ruff, so that doesn't say much for mypy's approach! I didn't realise that pyright also treats `Unbound` as a distinct type to `Unknown`.